File: src/java/org/homelending/appstore/workflow/WorkflowEngine.java
package org.homelending.appstore.workflow;

import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Java implementation of the Workflow Engine for the AI App Store
 * Provides functionality for creating, managing and executing workflows
 */
public class WorkflowEngine {
    private static final Logger logger = Logger.getLogger(WorkflowEngine.class.getName());
    
    private final Map<String, Workflow> workflows;
    private final Map<String, WorkflowExecution> executions;
    private final Map<String, Map<String, Function<Map<String, Object>, Map<String, Object>>>> componentRegistry;
    
    /**
     * Initialize the workflow engine
     */
    public WorkflowEngine() {
        this.workflows = new ConcurrentHashMap<>();
        this.executions = new ConcurrentHashMap<>();
        this.componentRegistry = new ConcurrentHashMap<>();
        logger.info("Workflow Engine initialized");
    }
    
    /**
     * Register a workflow with the engine
     * @param workflow The workflow to register
     * @return The ID of the registered workflow
     */
    public String registerWorkflow(Workflow workflow) {
        workflows.put(workflow.getId(), workflow);
        logger.info("Registered workflow: " + workflow.getId() + " - " + workflow.getName());
        return workflow.getId();
    }
    
    /**
     * Register a component with available actions
     * @param componentId ID of the component
     * @param actions Map of action names to functions
     */
    public void registerComponent(String componentId, 
                                 Map<String, Function<Map<String, Object>, Map<String, Object>>> actions) {
        componentRegistry.put(componentId, actions);
        logger.info("Registered component: " + componentId + " with actions: " + actions.keySet());
    }
    
    /**
     * Execute a workflow with the given inputs
     * @param workflowId ID of the workflow to execute
     * @param inputs Input data for the workflow
     * @return The workflow execution object
     */
    public WorkflowExecution executeWorkflow(String workflowId, Map<String, Object> inputs) {
        if (!workflows.containsKey(workflowId)) {
            throw new IllegalArgumentException("Workflow not found: " + workflowId);
        }
        
        Workflow workflow = workflows.get(workflowId);
        WorkflowExecution execution = new WorkflowExecution(workflow);
        executions.put(execution.getId(), execution);
        
        logger.info("Starting workflow execution: " + execution.getId() + " for workflow: " + workflowId);
        
        try {
            // Execute each step in sequence
            Map<String, Map<String, Object>> stepResults = new HashMap<>();
            
            for (int i = 0; i < workflow.getSteps().size(); i++) {
                WorkflowStep step = workflow.getSteps().get(i);
                logger.info("Executing step " + (i+1) + "/" + workflow.getSteps().size() + ": " + step.getName());
                
                // If component not registered, skip with error
                if (!componentRegistry.containsKey(step.getComponentId())) {
                    String errorMsg = "Component not registered: " + step.getComponentId();
                    Map<String, Object> error = new HashMap<>();
                    error.put("step", step.getId());
                    error.put("error", errorMsg);
                    execution.getErrors().add(error);
                    logger.severe(errorMsg);
                    continue;
                }
                
                Map<String, Function<Map<String, Object>, Map<String, Object>>> component = 
                    componentRegistry.get(step.getComponentId());
                
                // If action not registered, skip with error
                if (!component.containsKey(step.getAction())) {
                    String errorMsg = "Action not found: " + step.getAction() + " for component: " + step.getComponentId();
                    Map<String, Object> error = new HashMap<>();
                    error.put("step", step.getId());
                    error.put("error", errorMsg);
                    execution.getErrors().add(error);
                    logger.severe(errorMsg);
                    continue;
                }
                
                // Map inputs
                Map<String, Object> stepInputs = new HashMap<>();
                for (Map.Entry<String, String> mapping : step.getInputMapping().entrySet()) {
                    String target = mapping.getKey();
                    String source = mapping.getValue();
                    
                    if (source.startsWith("$.inputs.")) {
                        String inputKey = source.substring("$.inputs.".length());
                        if (inputs.containsKey(inputKey)) {
                            stepInputs.put(target, inputs.get(inputKey));
                        }
                    } else if (source.startsWith("$.steps.")) {
                        String[] parts = source.split("\\.", 4);
                        if (parts.length >= 4) {
                            String sourceStepId = parts[2];
                            String outputKey = parts[3];
                            
                            if (stepResults.containsKey(sourceStepId) && 
                                stepResults.get(sourceStepId).containsKey(outputKey)) {
                                stepInputs.put(target, stepResults.get(sourceStepId).get(outputKey));
                            }
                        }
                    }
                }
                
                // Execute action
                try {
                    Function<Map<String, Object>, Map<String, Object>> actionFunc = component.get(step.getAction());
                    Map<String, Object> result = actionFunc.apply(stepInputs);
                    stepResults.put(step.getId(), result);
                    
                    // Map outputs to workflow results
                    for (Map.Entry<String, String> mapping : step.getOutputMapping().entrySet()) {
                        String target = mapping.getKey();
                        String source = mapping.getValue();
                        
                        if (result.containsKey(source)) {
                            execution.getResults().put(target, result.get(source));
                        }
                    }
                } catch (Exception e) {
                    String errorMsg = "Error executing step " + step.getId() + ": " + e.getMessage();
                    Map<String, Object> error = new HashMap<>();
                    error.put("step", step.getId());
                    error.put("error", errorMsg);
                    execution.getErrors().add(error);
                    logger.log(Level.SEVERE, errorMsg, e);
                }
            }
            
            // Mark execution as complete
            boolean success = execution.getErrors().isEmpty();
            execution.complete(success);
            
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Error executing workflow: " + e.getMessage(), e);
            Map<String, Object> error = new HashMap<>();
            error.put("workflow", workflowId);
            error.put("error", e.getMessage());
            execution.getErrors().add(error);
            execution.complete(false);
        }
        
        return execution;
    }
    
    /**
     * Get the status of a workflow execution
     * @param executionId ID of the execution
     * @return Execution details
     */
    public Map<String, Object> getExecutionStatus(String executionId) {
        if (!executions.containsKey(executionId)) {
            throw new IllegalArgumentException("Execution not found: " + executionId);
        }
        
        WorkflowExecution execution = executions.get(executionId);
        return execution.toMap();
    }
    
    /**
     * Workflow step class
     */
    public static class WorkflowStep {
        private final String id;
        private final String name;
        private final String componentId;
        private final String action;
        private final Map<String, String> inputMapping;
        private final Map<String, String> outputMapping;
        
        /**
         * Create a workflow step
         * @param name Name of the step
         * @param componentId ID of the component to use
         * @param action Action to perform
         * @param inputMapping Mapping of inputs
         * @param outputMapping Mapping of outputs
         */
        public WorkflowStep(String name, String componentId, String action, 
                           Map<String, String> inputMapping, Map<String, String> outputMapping) {
            this.id = UUID.randomUUID().toString();
            this.name = name;
            this.componentId = componentId;
            this.action = action;
            this.inputMapping = inputMapping != null ? inputMapping : new HashMap<>();
            this.outputMapping = outputMapping != null ? outputMapping : new HashMap<>();
        }
        
        public String getId() { return id; }
        public String getName() { return name; }
        public String getComponentId() { return componentId; }
        public String getAction() { return action; }
        public Map<String, String> getInputMapping() { return inputMapping; }
        public Map<String, String> getOutputMapping() { return outputMapping; }
        
        /**
         * Convert to a map representation
         * @return Map representation of the step
         */
        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("id", id);
            map.put("name", name);
            map.put("component_id", componentId);
            map.put("action", action);
            map.put("input_mapping", inputMapping);
            map.put("output_mapping", outputMapping);
            return map;
        }
    }
    
    /**
     * Workflow class
     */
    public static class Workflow {
        private final String id;
        private final String name;
        private final String description;
        private final List<WorkflowStep> steps;
        private final LocalDateTime createdAt;
        private LocalDateTime updatedAt;
        private String status;
        
        /**
         * Create a workflow
         * @param name Name of the workflow
         * @param description Description of the workflow
         */
        public Workflow(String name, String description) {
            this.id = UUID.randomUUID().toString();
            this.name = name;
            this.description = description;
            this.steps = new ArrayList<>();
            this.createdAt = LocalDateTime.now();
            this.updatedAt = this.createdAt;
            this.status = "draft";
        }
        
        public String getId() { return id; }
        public String getName() { return name; }
        public String getDescription() { return description; }
        public List<WorkflowStep> getSteps() { return steps; }
        public LocalDateTime getCreatedAt() { return createdAt; }
        public LocalDateTime getUpdatedAt() { return updatedAt; }
        public String getStatus() { return status; }
        
        /**
         * Add a step to the workflow
         * @param step Step to add
         */
        public void addStep(WorkflowStep step) {
            steps.add(step);
            updatedAt = LocalDateTime.now();
        }
        
        /**
         * Convert to a map representation
         * @return Map representation of the workflow
         */
        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("id", id);
            map.put("name", name);
            map.put("description", description);
            
            List<Map<String, Object>> stepsList = new ArrayList<>();
            for (WorkflowStep step : steps) {
                stepsList.add(step.toMap());
            }
            map.put("steps", stepsList);
            
            map.put("created_at", createdAt.toString());
            map.put("updated_at", updatedAt.toString());
            map.put("status", status);
            
            return map;
        }
    }
    
    /**
     * Workflow execution class
     */
    public static class WorkflowExecution {
        private final String id;
        private final String workflowId;
        private final Workflow workflow;
        private final LocalDateTime startedAt;
        private LocalDateTime completedAt;
        private String status;
        private final Map<String, Object> results;
        private final List<Map<String, Object>> errors;
        
        /**
         * Create a workflow execution
         * @param workflow Workflow to execute
         */
        public WorkflowExecution(Workflow workflow) {
            this.id = UUID.randomUUID().toString();
            this.workflowId = workflow.getId();
            this.workflow = workflow;
            this.startedAt = LocalDateTime.now();
            this.status = "running";
            this.results = new HashMap<>();
            this.errors = new ArrayList<>();
        }
        
        public String getId() { return id; }
        public String getWorkflowId() { return workflowId; }
        public Workflow getWorkflow() { return workflow; }
        public LocalDateTime getStartedAt() { return startedAt; }
        public LocalDateTime getCompletedAt() { return completedAt; }
        public String getStatus() { return status; }
        public Map<String, Object> getResults() { return results; }
        public List<Map<String, Object>> getErrors() { return errors; }
        
        /**
         * Mark the execution as complete
         * @param success Whether the execution was successful
         */
        public void complete(boolean success) {
            this.completedAt = LocalDateTime.now();
            this.status = success ? "completed" : "failed";
        }
        
        /**
         * Convert to a map representation
         * @return Map representation of the execution
         */
        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("id", id);
            map.put("workflow_id", workflowId);
            map.put("started_at", startedAt.toString());
            map.put("completed_at", completedAt != null ? completedAt.toString() : null);
            map.put("status", status);
            map.put("results", results);
            map.put("errors", errors);
            return map;
        }
    }
    
    /**
     * Main method for demonstration
     */
    public static void main(String[] args) {
        // Create workflow engine
        WorkflowEngine engine = new WorkflowEngine();
        
        // Register a sample component
        Map<String, Function<Map<String, Object>, Map<String, Object>>> dataProcessor = new HashMap<>();
        
        // Sample action: parse JSON data
        dataProcessor.put("parse_json", inputs -> {
            Map<String, Object> result = new HashMap<>();
            String jsonData = (String) inputs.get("json_data");
            // In a real implementation, this would parse the JSON
            result.put("result", "Parsed " + jsonData);
            return result;
        });
        
        engine.registerComponent("data_processor", dataProcessor);
        
        // Create a sample workflow
        Workflow workflow = new Workflow("Sample Workflow", "Demonstration workflow");
        
        // Add a step
        Map<String, String> inputMapping = new HashMap<>();
        inputMapping.put("json_data", "$.inputs.data");
        
        Map<String, String> outputMapping = new HashMap<>();
        outputMapping.put("parsed_result", "result");
        
        WorkflowStep step = new WorkflowStep(
            "Parse Data", 
            "data_processor", 
            "parse_json", 
            inputMapping, 
            outputMapping
        );
        
        workflow.addStep(step);
        
        // Register the workflow
        engine.registerWorkflow(workflow);
        
        // Execute the workflow with sample input
        Map<String, Object> inputs = new HashMap<>();
        inputs.put("data", "{\"key\": \"value\"}");
        
        WorkflowExecution execution = engine.executeWorkflow(workflow.getId(), inputs);
        
        // Print execution results
        System.out.println("Execution ID: " + execution.getId());
        System.out.println("Status: " + execution.getStatus());
        System.out.println("Results: " + execution.getResults());
    }
}

File: src/python/workflows/workflow_engine.py
#!/usr/bin/env python3
"""
Workflow Engine for the AI App Store

This module provides functionality for creating, managing, and executing
workflows that chain together multiple AI applications.
"""
import json
import logging
import uuid
from datetime import datetime
from typing import Dict, List, Any, Optional, Callable

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class WorkflowStep:
    """Represents a single step in a workflow"""
    
    def __init__(self, 
                 name: str, 
                 component_id: str, 
                 action: str, 
                 input_mapping: Dict[str, str] = None,
                 output_mapping: Dict[str, str] = None):
        """
        Initialize a workflow step
        
        Args:
            name: Human-readable name for the step
            component_id: ID of the component/app to execute
            action: Action to perform on the component
            input_mapping: Maps workflow inputs to component inputs
            output_mapping: Maps component outputs to workflow outputs
        """
        self.id = str(uuid.uuid4())
        self.name = name
        self.component_id = component_id
        self.action = action
        self.input_mapping = input_mapping or {}
        self.output_mapping = output_mapping or {}
        
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary representation"""
        return {
            "id": self.id,
            "name": self.name,
            "component_id": self.component_id,
            "action": self.action,
            "input_mapping": self.input_mapping,
            "output_mapping": self.output_mapping
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'WorkflowStep':
        """Create from dictionary representation"""
        step = cls(
            name=data["name"],
            component_id=data["component_id"],
            action=data["action"],
            input_mapping=data.get("input_mapping", {}),
            output_mapping=data.get("output_mapping", {})
        )
        step.id = data["id"]
        return step
        

class Workflow:
    """Represents a workflow that chains together multiple applications"""
    
    def __init__(self, name: str, description: str = ""):
        """
        Initialize a workflow
        
        Args:
            name: Name of the workflow
            description: Optional description
        """
        self.id = str(uuid.uuid4())
        self.name = name
        self.description = description
        self.steps: List[WorkflowStep] = []
        self.created_at = datetime.utcnow()
        self.updated_at = self.created_at
        self.status = "draft"
        
    def add_step(self, step: WorkflowStep) -> None:
        """Add a step to the workflow"""
        self.steps.append(step)
        self.updated_at = datetime.utcnow()
        
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary representation"""
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "steps": [step.to_dict() for step in self.steps],
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "status": self.status
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Workflow':
        """Create from dictionary representation"""
        workflow = cls(
            name=data["name"],
            description=data.get("description", "")
        )
        workflow.id = data["id"]
        workflow.status = data.get("status", "draft")
        workflow.created_at = datetime.fromisoformat(data["created_at"])
        workflow.updated_at = datetime.fromisoformat(data["updated_at"])
        workflow.steps = [WorkflowStep.from_dict(step) for step in data["steps"]]
        return workflow
    
    def save(self, filepath: str) -> None:
        """Save workflow to file"""
        with open(filepath, 'w') as f:
            json.dump(self.to_dict(), f, indent=2)
    
    @classmethod
    def load(cls, filepath: str) -> 'Workflow':
        """Load workflow from file"""
        with open(filepath, 'r') as f:
            data = json.load(f)
        return cls.from_dict(data)


class WorkflowExecution:
    """Represents a single execution of a workflow"""
    
    def __init__(self, workflow: Workflow):
        """
        Initialize a workflow execution
        
        Args:
            workflow: The workflow to execute
        """
        self.id = str(uuid.uuid4())
        self.workflow_id = workflow.id
        self.workflow = workflow
        self.started_at = datetime.utcnow()
        self.completed_at = None
        self.status = "running"
        self.results = {}
        self.errors = []
        
    def complete(self, success: bool) -> None:
        """Mark the execution as complete"""
        self.completed_at = datetime.utcnow()
        self.status = "completed" if success else "failed"
        
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary representation"""
        return {
            "id": self.id,
            "workflow_id": self.workflow_id,
            "started_at": self.started_at.isoformat(),
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "status": self.status,
            "results": self.results,
            "errors": self.errors
        }


class WorkflowEngine:
    """Engine for executing workflows"""
    
    def __init__(self):
        """Initialize the workflow engine"""
        self.workflows: Dict[str, Workflow] = {}
        self.executions: Dict[str, WorkflowExecution] = {}
        self.component_registry: Dict[str, Dict[str, Callable]] = {}
        
    def register_workflow(self, workflow: Workflow) -> str:
        """Register a workflow with the engine"""
        self.workflows[workflow.id] = workflow
        logger.info(f"Registered workflow: {workflow.id} - {workflow.name}")
        return workflow.id
    
    def register_component(self, 
                          component_id: str, 
                          actions: Dict[str, Callable]) -> None:
        """
        Register a component with available actions
        
        Args:
            component_id: ID of the component
            actions: Dictionary of action names to functions
        """
        self.component_registry[component_id] = actions
        logger.info(f"Registered component: {component_id} with actions: {list(actions.keys())}")
    
    def execute_workflow(self, 
                        workflow_id: str, 
                        inputs: Dict[str, Any]) -> WorkflowExecution:
        """
        Execute a workflow with the given inputs
        
        Args:
            workflow_id: ID of the workflow to execute
            inputs: Input data for the workflow
            
        Returns:
            The workflow execution object
        """
        if workflow_id not in self.workflows:
            raise ValueError(f"Workflow not found: {workflow_id}")
            
        workflow = self.workflows[workflow_id]
        execution = WorkflowExecution(workflow)
        self.executions[execution.id] = execution
        
        logger.info(f"Starting workflow execution: {execution.id} for workflow: {workflow_id}")
        
        try:
            # Execute each step in sequence
            step_results = {}
            for i, step in enumerate(workflow.steps):
                logger.info(f"Executing step {i+1}/{len(workflow.steps)}: {step.name}")
                
                # If component not registered, skip with error
                if step.component_id not in self.component_registry:
                    error_msg = f"Component not registered: {step.component_id}"
                    execution.errors.append({
                        "step": step.id,
                        "error": error_msg
                    })
                    logger.error(error_msg)
                    continue
                    
                component = self.component_registry[step.component_id]
                
                # If action not registered, skip with error
                if step.action not in component:
                    error_msg = f"Action not found: {step.action} for component: {step.component_id}"
                    execution.errors.append({
                        "step": step.id,
                        "error": error_msg
                    })
                    logger.error(error_msg)
                    continue
                
                # Map inputs
                step_inputs = {}
                for target, source in step.input_mapping.items():
                    if source.startswith("$.inputs."):
                        input_key = source[len("$.inputs."):]
                        if input_key in inputs:
                            step_inputs[target] = inputs[input_key]
                    elif source.startswith("$.steps."):
                        parts = source.split(".")
                        if len(parts) >= 3:
                            source_step_id = parts[2]
                            output_key = ".".join(parts[3:])
                            if (source_step_id in step_results and 
                                output_key in step_results[source_step_id]):
                                step_inputs[target] = step_results[source_step_id][output_key]
                
                # Execute action
                try:
                    action_func = component[step.action]
                    result = action_func(**step_inputs)
                    step_results[step.id] = result
                    
                    # Map outputs to workflow results
                    for target, source in step.output_mapping.items():
                        if source in result:
                            execution.results[target] = result[source]
                            
                except Exception as e:
                    error_msg = f"Error executing step {step.id}: {str(e)}"
                    execution.errors.append({
                        "step": step.id,
                        "error": error_msg
                    })
                    logger.error(error_msg, exc_info=True)
            
            # Mark execution as complete
            success = len(execution.errors) == 0
            execution.complete(success)
            
        except Exception as e:
            logger.error(f"Error executing workflow: {str(e)}", exc_info=True)
            execution.errors.append({
                "workflow": workflow_id,
                "error": str(e)
            })
            execution.complete(False)
            
        return execution
    
    def get_execution_status(self, execution_id: str) -> Dict[str, Any]:
        """Get the status of a workflow execution"""
        if execution_id not in self.executions:
            raise ValueError(f"Execution not found: {execution_id}")
            
        execution = self.executions[execution_id]
        return execution.to_dict()


# Example usage
if __name__ == "__main__":
    # Create a simple workflow
    workflow = Workflow("Example Workflow", "A simple example workflow")
    
    # Add steps
    step1 = WorkflowStep(
        name="Parse Input",
        component_id="data_processor",
        action="parse_json",
        input_mapping={"json_data": "$.inputs.data"},
        output_mapping={"parsed_data": "result"}
    )
    workflow.add_step(step1)
    
    step2 = WorkflowStep(
        name="Analyze Data",
        component_id="ai_analyzer",
        action="analyze",
        input_mapping={"data": "$.steps." + step1.id + ".result"},
        output_mapping={"analysis_result": "analysis"}
    )
    workflow.add_step(step2)
    
    # Print workflow definition
    print(json.dumps(workflow.to_dict(), indent=2))